README.md
<p align=center>
  <a href="https://github.com/TimoKunze/TabStripControl/releases/tag/1.10.5">
    <img alt="Release 1.10.5 Release" src="https://img.shields.io/badge/release-1.10.5-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/TabStripControl/releases">
    <img alt="Download TabStripControl" src="https://img.shields.io/badge/download-latest-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/TabStripControl/blob/master/LICENSE">
    <img alt="License" src="https://img.shields.io/badge/license-MIT-0688CB.svg">
  </a>
  <a href="https://www.paypal.com/xclick/business=TKunze71216%40gmx.de&item_name=TabStripControl&no_shipping=1&tax=0&currency_code=EUR">
    <img alt="Donate" src="https://img.shields.io/badge/%24-donate-E44E4A.svg">
  </a>
</p>

# TabStripControl
An ActiveX control for Visual Basic 6 that can be used to organize UI elements in tabs.

I've developed this ActiveX control in 2006 and did update it on a regular basis until 2016. Currently I have little interest to maintain this project any longer, but I think the code might be of some use to others.

# Before you make changes
If you make changes to the code and deploy the binary, keep in mind that ActiveX controls are COM components and therefore should stay binary compatible as long as you don't change the COM object's, i.e. the ActiveX control's public class name and GUIDs. Otherwise people using these components are likely to end up in the famous COM hell.

# Dependencies
You will need the [Microsoft Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk), ATL, and [WTL 10](https://sourceforge.net/projects/wtl/).

## Required Change to ATL
Some version of ATL have a bug in ```AtlIPersistPropertyBag_Load``` which causes crashes. In the file atlcom.h search for ```AtlIPersistPropertyBag_Load```. Inside the implementation of this function search for ```HRESULT hr = pPropBag->Read(pMap[i].szDesc, &var, pErrorLog);```. Make sure that there is this code before this line: ```var.pdispVal = NULL;```.

CHANGELOG
Version 1.10.5 - 28/04/2019

- FIX: Dragging data to certain programs like Total Commander or FileZilla did not work if the data was provided only in the OLESetData event.



Version 1.10.4 - 04/02/2018

- Upgraded to WTL 10.
- Integrated the GitHub repository into the documentation and the About dialog.



Version 1.10.3 - 16/01/2018

- Project published on GitHub.



Version 1.10.2 - 31/07/2016

- Fixed some memory leaks that occurred when firing certain events.



Version 1.10.1 - 26/01/2016

- FIX: The control could crash during OLE drag'n'drop.
- FIX: The control could crash the other process during OLE drag'n'drop, especially when IOLEDataObject::SetDropDescription() had been called.



Version 1.10.0 - 23/01/2016

- Added the ITabStrip::FinishOLEDragDrop method. It should not be used if not really needed. It can be used to hide the drag image if execution of the OLEDragDrop event takes a long time or if a message is displayed from within this event.



Version 1.9.0 - 13/09/2015

- Added the ITabStripTab::HasCloseButton property.
- Initial support for Windows 10.



Version 1.8.0 - 20/07/2014

- Added the XClick, and XDblClick events for support of the extended mouse buttons.
- Extended the Mouse* events to support the extended mouse buttons.
- FIX: The Font and MouseIcon properties could leak GDI resources.
- Initial support for Windows 8.1.



Version 1.7.0 - 21/09/2013

- Added the ITabStrip::CloseableTabsMode property.



Version 1.6.0 - 05/04/2013

- Performance improvements.



Version 1.5.2 - 13/01/2013

- FIX: The last character of the tab text could not be used as accelerator key.
- FIX: The ResizedControlWindow event was not fired when it should have been fired.
- FIX: Drawing issues could occur if a windowless control was placed inside the control.
- Initial support for Windows 8.



Version 1.5.1 - 02/03/2012

- Fixed a memory leak that occurred during drag'n'drop if Aero drag-images have been used.
- Reimplemented the subclassing code of the VB6 samples to make it more robust.



Version 1.5.0 - 01/11/2011

- Added the ITabStripTab::hAssociatedWindow property.



Version 1.4.1 - 21/05/2011

- FIX: Disabling the control did not move the keyboard focus to the next control in the tab order.
- FIX: The ResizedControlWindow event was fired when actually nothing had changed.



Version 1.4.0 - 08/01/2011

- FIX: The control could crash when it was destroyed.
- FIX: Due do a bug in Microsoft ATL, the control could lose keyboard focus when it shouldn't. This problem occured in MDI applications only. Many thanks to Peter Schellenbach for his help.
- Switched to Visual C++ 2010 and dropped support for Windows 2000.



Version 1.3.2 - 07/11/2010

- FIX: IOLEDataObject::SetDropDescription could cause crashes in other programs.



Version 1.3.1 - 02/01/2010

- Fixed a potential problem with persistance.
- Improved support for Aero drag images.
- FIX: If the control was placed on a Form for which the KeyPreview feature was activated, the Form_Key* handler didn't receive key strokes like all Alt+PgDown or Alt+Left, if the control had the keyboard focus.
- FIX: The 'shift' parameter that many events have, often missed the ALT key.
- Deploy (stripped) symbols with the binaries to make debugging easier. This are the pdb files in case you wonder.



Version 1.3.0 - 11/07/2009

- Added the ITabStrip::OLEDragImageStyle property.
- FIX: When using closeable tabs, scrolling tabs could leave some debris around the scroll buttons.
- Massively reduced the usage of GDI handles.
- It is no longer necessary to subclass the control's parent window and reflect the WM_CONTEXTMENU message to make the ContextMenu event work.
- Made some internal changes that will make it easier for me to keep binary compatibility in future.



Version 1.2.1 - 18/02/2009

- Initial support for Windows 7.
- Extended the DropDescriptionIconConstants enumeration by 'ddiUseDefault'.
- Fixed some small OLE drag'n'drop bugs.
- FIX: The TAB key didn't move the focus to the next control if the control was placed within a VB6 UserControl.
- FIX: Changing the RightToLeft property at runtime did not fully work if rtlLayout was involved.



Version 1.2.0 - 31/10/2008

- Added the _ITabStripEvents::OLEDragEnterPotentialTarget event.
- Added the _ITabStripEvents::OLEDragLeavePotentialTarget event.
- Added the _ITabStripEvents::OLEReceivedNewData event.
- FIX: The _ITabStripEvents::FreeTabData event didn't work as expected. The provided tab was always Nothing.
- FIX: The _ITabStripEvents::MouseLeave event didn't work as expected. It was raised everytime the mouse wheel was scrolled outside the control if the control had the focus.
- FIX: Ignore calls to IOLEDataObject::SetDropDescription() if the drag image is in classic style.
- FIX: It wasn't possible to change the active tab using [CTRL]+[TAB] and [SHIFT]+[CTRL]+[TAB].
- Fixed some GDI handle leaks.
- Fixed some drawing glitches.
- Removed the data-binding stuff (it was probably broken anyway).
- Switched to Visual C++ 2008 and dropped support for Windows 9x, Me and NT 4.0.
- Improved the documentation.



Version 1.1.0 - 17/12/2007

- Added the ITabStrip::hHighResImageList property.
- Added the ITabStripTab::CreateDragImage() method.
- Added the ITabStripTabContainer::CreateDragImage() method.
- Added the IOLEDataObject::GetDropDescription() method.
- Added the IOLEDataObject::SetDropDescription() method.
- Extended ITabStrip::OLEDrag() by the parameter 'itemCountToDisplay'.
- Implemented built-in drag-image creation.
- Rewrote large parts of the drag'n'drop code.
- FIX: Keyboard input wasn't processed if the control was placed within a VB6 UserControl.
- FIX: On Windows 98/ME tooltips didn't work.
- FIX: Changes were not applied implicitly when jumping between multiple instances of the control with the property pages opened.
- Improved the documentation.



Version 1.0.4 - 28/02/2007

- Decreased the binaries' size.
- Improved the documentation.

src/DispIDs.h
// DispIDs.h: Defines a DISPID for each COM property and method we're providing.

// ITabStrip
// properties
#define DISPID_TABSTRIPCTL_ACTIVETAB											DISPID_VALUE
#define DISPID_TABSTRIPCTL_ALLOWDRAGDROP									1
#define DISPID_TABSTRIPCTL_APPEARANCE											2
#define DISPID_TABSTRIPCTL_BORDERSTYLE										3
#define DISPID_TABSTRIPCTL_CARETTAB												4
#define DISPID_TABSTRIPCTL_CLOSEABLETABS									5
#define DISPID_TABSTRIPCTL_DISABLEDEVENTS									6
#define DISPID_TABSTRIPCTL_DISPLAYAREAHEIGHT							7
#define DISPID_TABSTRIPCTL_DISPLAYAREALEFT								8
#define DISPID_TABSTRIPCTL_DISPLAYAREATOP									9
#define DISPID_TABSTRIPCTL_DISPLAYAREAWIDTH								10
//#define DISPID_TABSTRIPCTL_DONTREDRAW											11
#define DISPID_TABSTRIPCTL_DRAGACTIVATETIME								12
#define DISPID_TABSTRIPCTL_DRAGGEDTABS										13
#define DISPID_TABSTRIPCTL_DRAGSCROLLTIMEBASE							14
#define DISPID_TABSTRIPCTL_DROPHILITEDTAB									15
#define DISPID_TABSTRIPCTL_ENABLED												DISPID_ENABLED
#define DISPID_TABSTRIPCTL_FIXEDTABWIDTH									16
#define DISPID_TABSTRIPCTL_FOCUSONBUTTONDOWN							17
#define DISPID_TABSTRIPCTL_FONT														18
#define DISPID_TABSTRIPCTL_HDRAGIMAGELIST									19
#define DISPID_TABSTRIPCTL_HIMAGELIST											20
#define DISPID_TABSTRIPCTL_HORIZONTALPADDING							21
#define DISPID_TABSTRIPCTL_HOTTRACKING										22
#define DISPID_TABSTRIPCTL_HOVERTIME											23
#define DISPID_TABSTRIPCTL_HWND														24
#define DISPID_TABSTRIPCTL_HWNDARROWBUTTONS								25
#define DISPID_TABSTRIPCTL_HWNDTOOLTIP										26
#define DISPID_TABSTRIPCTL_INSERTMARKCOLOR								27
#define DISPID_TABSTRIPCTL_MINTABWIDTH										28
#define DISPID_TABSTRIPCTL_MOUSEICON											29
#define DISPID_TABSTRIPCTL_MOUSEPOINTER										30
#define DISPID_TABSTRIPCTL_MULTIROW												31
#define DISPID_TABSTRIPCTL_MULTISELECT										32
#define DISPID_TABSTRIPCTL_OWNERDRAWN											33
#define DISPID_TABSTRIPCTL_PROCESSCONTEXTMENUKEYS					34
#define DISPID_TABSTRIPCTL_RAGGEDTABROWS									35
#define DISPID_TABSTRIPCTL_REGISTERFOROLEDRAGDROP					36
#define DISPID_TABSTRIPCTL_RIGHTTOLEFT										37
#define DISPID_TABSTRIPCTL_SCROLLTOOPPOSITE								38
#define DISPID_TABSTRIPCTL_SHOWBUTTONSEPARATORS						39
#define DISPID_TABSTRIPCTL_SHOWDRAGIMAGE									40
#define DISPID_TABSTRIPCTL_SHOWTOOLTIPS										41
#define DISPID_TABSTRIPCTL_STYLE													42
#define DISPID_TABSTRIPCTL_SUPPORTOLEDRAGIMAGES						43
#define DISPID_TABSTRIPCTL_TABBOUNDINGBOXDEFINITION				44
#define DISPID_TABSTRIPCTL_TABCAPTIONALIGNMENT						45
#define DISPID_TABSTRIPCTL_TABHEIGHT											46
#define DISPID_TABSTRIPCTL_TABPLACEMENT										47
#define DISPID_TABSTRIPCTL_TABS														48
#define DISPID_TABSTRIPCTL_USEFIXEDTABWIDTH								49
#define DISPID_TABSTRIPCTL_USESYSTEMFONT									50
#define DISPID_TABSTRIPCTL_VERSION												51
#define DISPID_TABSTRIPCTL_VERTICALPADDING								52
#define DISPID_TABSTRIPCTL_HHIGHRESIMAGELIST							66
#define DISPID_TABSTRIPCTL_OLEDRAGIMAGESTYLE							67
//#define DISPID_TABSTRIPCTL_REFLECTCONTEXTMENUMESSAGES			68
#define DISPID_TABSTRIPCTL_CLOSEABLETABSMODE							69
// fingerprint
#define DISPID_TABSTRIPCTL_APPID													500
#define DISPID_TABSTRIPCTL_APPNAME												501
#define DISPID_TABSTRIPCTL_APPSHORTNAME										502
#define DISPID_TABSTRIPCTL_BUILD													503
#define DISPID_TABSTRIPCTL_CHARSET												504
#define DISPID_TABSTRIPCTL_ISRELEASE											505
#define DISPID_TABSTRIPCTL_PROGRAMMER											506
#define DISPID_TABSTRIPCTL_TESTER													507
// methods
#define DISPID_TABSTRIPCTL_ABOUT													DISPID_ABOUTBOX
#define DISPID_TABSTRIPCTL_BEGINDRAG											53
#define DISPID_TABSTRIPCTL_CALCULATEDISPLAYAREA						54
#define DISPID_TABSTRIPCTL_CALCULATEWINDOWRECTANGLE				55
#define DISPID_TABSTRIPCTL_COUNTTABROWS										56
#define DISPID_TABSTRIPCTL_CREATETABCONTAINER							57
#define DISPID_TABSTRIPCTL_ENDDRAG												58
#define DISPID_TABSTRIPCTL_GETCLOSESTINSERTMARKPOSITION		59
#define DISPID_TABSTRIPCTL_GETINSERTMARKPOSITION					60
#define DISPID_TABSTRIPCTL_HITTEST												61
#define DISPID_TABSTRIPCTL_LOADSETTINGSFROMFILE						62
#define DISPID_TABSTRIPCTL_OLEDRAG												63
#define DISPID_TABSTRIPCTL_REFRESH												DISPID_REFRESH
#define DISPID_TABSTRIPCTL_SAVESETTINGSTOFILE							64
#define DISPID_TABSTRIPCTL_SETINSERTMARKPOSITION					65
#define DISPID_TABSTRIPCTL_FINISHOLEDRAGDROP							70


// _ITabStripEvents
// methods
#define DISPID_TABSTRIPCTLE_ABORTEDDRAG										1
#define DISPID_TABSTRIPCTLE_ACTIVETABCHANGED							DISPID_VALUE
#define DISPID_TABSTRIPCTLE_ACTIVETABCHANGING							2
#define DISPID_TABSTRIPCTLE_CARETCHANGED									3
#define DISPID_TABSTRIPCTLE_CLICK													4
#define DISPID_TABSTRIPCTLE_CONTEXTMENU										5
#define DISPID_TABSTRIPCTLE_DBLCLICK											6
#define DISPID_TABSTRIPCTLE_DESTROYEDCONTROLWINDOW				7
#define DISPID_TABSTRIPCTLE_DRAGMOUSEMOVE									8
#define DISPID_TABSTRIPCTLE_DROP													9
#define DISPID_TABSTRIPCTLE_FREETABDATA										10
#define DISPID_TABSTRIPCTLE_INSERTEDTAB										11
#define DISPID_TABSTRIPCTLE_INSERTINGTAB									12
#define DISPID_TABSTRIPCTLE_KEYDOWN												13
#define DISPID_TABSTRIPCTLE_KEYPRESS											14
#define DISPID_TABSTRIPCTLE_KEYUP													15
#define DISPID_TABSTRIPCTLE_MCLICK												16
#define DISPID_TABSTRIPCTLE_MDBLCLICK											17
#define DISPID_TABSTRIPCTLE_MOUSEDOWN											18
#define DISPID_TABSTRIPCTLE_MOUSEENTER										19
#define DISPID_TABSTRIPCTLE_MOUSEHOVER										20
#define DISPID_TABSTRIPCTLE_MOUSELEAVE										21
#define DISPID_TABSTRIPCTLE_MOUSEMOVE											22
#define DISPID_TABSTRIPCTLE_MOUSEUP												23
#define DISPID_TABSTRIPCTLE_OLECOMPLETEDRAG								24
#define DISPID_TABSTRIPCTLE_OLEDRAGDROP										25
#define DISPID_TABSTRIPCTLE_OLEDRAGENTER									26
#define DISPID_TABSTRIPCTLE_OLEDRAGLEAVE									27
#define DISPID_TABSTRIPCTLE_OLEDRAGMOUSEMOVE							28
#define DISPID_TABSTRIPCTLE_OLEGIVEFEEDBACK								29
#define DISPID_TABSTRIPCTLE_OLEQUERYCONTINUEDRAG					30
#define DISPID_TABSTRIPCTLE_OLESETDATA										31
#define DISPID_TABSTRIPCTLE_OLESTARTDRAG									32
#define DISPID_TABSTRIPCTLE_OWNERDRAWTAB									33
#define DISPID_TABSTRIPCTLE_RCLICK												34
#define DISPID_TABSTRIPCTLE_RDBLCLICK											35
#define DISPID_TABSTRIPCTLE_RECREATEDCONTROLWINDOW				36
#define DISPID_TABSTRIPCTLE_REMOVEDTAB										37
#define DISPID_TABSTRIPCTLE_REMOVINGTAB										38
#define DISPID_TABSTRIPCTLE_RESIZEDCONTROLWINDOW					39
#define DISPID_TABSTRIPCTLE_TABBEGINDRAG									40
#define DISPID_TABSTRIPCTLE_TABBEGINRDRAG									41
#define DISPID_TABSTRIPCTLE_TABGETINFOTIPTEXT							42
#define DISPID_TABSTRIPCTLE_TABMOUSEENTER									43
#define DISPID_TABSTRIPCTLE_TABMOUSELEAVE									44
#define DISPID_TABSTRIPCTLE_TABSELECTIONCHANGED						45
#define DISPID_TABSTRIPCTLE_OLEDRAGENTERPOTENTIALTARGET		46
#define DISPID_TABSTRIPCTLE_OLEDRAGLEAVEPOTENTIALTARGET		47
#define DISPID_TABSTRIPCTLE_OLERECEIVEDNEWDATA						48
#define DISPID_TABSTRIPCTLE_XCLICK												49
#define DISPID_TABSTRIPCTLE_XDBLCLICK											50


// ITabStripTab
// properties
#define DISPID_TST_ACTIVE																	1
#define DISPID_TST_CARET																	2
#define DISPID_TST_DROPHILITED														3
#define DISPID_TST_HEIGHT																	4
#define DISPID_TST_ICONINDEX															5
#define DISPID_TST_ID																			6
#define DISPID_TST_INDEX																	7
#define DISPID_TST_LEFT																		8
#define DISPID_TST_SELECTED																9
#define DISPID_TST_TABDATA																10
#define DISPID_TST_TEXT																		DISPID_VALUE
#define DISPID_TST_TOP																		11
#define DISPID_TST_WIDTH																	12
#define DISPID_TST_HASSOCIATEDWINDOW											14
#define DISPID_TST_HASCLOSEBUTTON													15
// methods
#define DISPID_TST_CREATEDRAGIMAGE												13


// ITabStripTabs
// properties
#define DISPID_TSTS_CASESENSITIVEFILTERS									1
#define DISPID_TSTS_COMPARISONFUNCTION										2
#define DISPID_TSTS_FILTER																3
#define DISPID_TSTS_FILTERTYPE														4
#define DISPID_TSTS_ITEM																	DISPID_VALUE
#define DISPID_TSTS__NEWENUM															DISPID_NEWENUM
// methods
#define DISPID_TSTS_ADD																		5
#define DISPID_TSTS_CONTAINS															6
#define DISPID_TSTS_COUNT																	7
#define DISPID_TSTS_DESELECTALL														8
#define DISPID_TSTS_REMOVE																9
#define DISPID_TSTS_REMOVEALL															10


// ITabStripTabContainer
// properties
#define DISPID_TSTC_ITEM																	DISPID_VALUE
#define DISPID_TSTC__NEWENUM															DISPID_NEWENUM
// methods
#define DISPID_TSTC_ADD																		1
#define DISPID_TSTC_CLONE																	2
#define DISPID_TSTC_COUNT																	3
#define DISPID_TSTC_CREATEDRAGIMAGE												4
#define DISPID_TSTC_REMOVE																5
#define DISPID_TSTC_REMOVEALL															6


// IVirtualTabStripTab
// properties
#define DISPID_VTST_DROPHILITED														1
#define DISPID_VTST_ICONINDEX															2
#define DISPID_VTST_INDEX																	3
#define DISPID_VTST_SELECTED															4
#define DISPID_VTST_TABDATA																5
#define DISPID_VTST_TEXT																	DISPID_VALUE
// methods


// IOLEDataObject
// properties
// methods
#define DISPID_ODO_CLEAR																	1
#define DISPID_ODO_GETCANONICALFORMAT											2
#define DISPID_ODO_GETDATA																3
#define DISPID_ODO_GETFORMAT															4
#define DISPID_ODO_SETDATA																5
#define DISPID_ODO_GETDROPDESCRIPTION											6
#define DISPID_ODO_SETDROPDESCRIPTION											7
